Automattic/wp-calypso
# CPT: Add trash button to post type list item (v2)

~~Cherry-picks from #5127 (00d4f315202c2e02f346cf8abc300f98ac430efe, 95d267d851dfeb5912dd2e49ef5e2edf5821d5f8)~~
Cherry-picks from #5007 (b0b0154c3ec2ab5f5f9e2f547622a98257880506)
Cherry-picks from #4557 (32a72102734e159d01f926df52930deca81cfd19, fdfdff29c67b3da6a533e25b9c408d714e93c774)

This pull request seeks to try an experimental alternative approach for dealing with post query state (and queryable state data generally) with better accommodation for updates to state over time. The included `QueryManager` class is intended to be extended by any use-case where we may need to track queried items and match, merge, and sort as revisions to that data are made in the application. The specific use-case in this pull request is in trashing a single post item. Refer to the "Open Questions" section of #4557 for more background on the complexities of this seemingly straight-forward interaction.

It's anticipated that if successful, this could be applied to terms state, which has similar requirements for filtering terms by search.

__Testing instructions:__

Ensure Mocha tests pass:

```
npm run test-client
```

Verify that you can trash posts whilst on the custom post type listing page. The post should instantly disappear from the post list when trashed.

For simplicity's sake, you can access the custom post types listing screen for the built-in posts type: http://calypso.localhost:3000/types/post

More interestingly, you can try variations where multiple queries are tracked simultaneously, noting that changing a post (e.g. trashing) will cause the item to appear in the newly applicable query. For example, navigating from "Trashed" filter to "Published", then trashing a post, if you were to navigate back to "Trashed", the new item should be reflected in the list with correct sorting, even before the REST API request has finished.

__Caveats:__

- There is not yet a notice to confirm that the post had been trashed successfully. 
- Post counts on the filter tabs do not yet update in accordance with changing status of post.

__Unknowns:__

- The continued existence of serialized query keys in the posts state reducer is somewhat undesirable (query request status, query last page). If these are necessary in other queryable state, we may want to seek a common approach.
- `queriesLastPage` post state would be affected by trashing posts, which is not reflected here
